xword_dl/downloader/avcxdownloader.py

```diff
@@ -0,0 +1,234 @@
+                "Could not find puzzle for specified date.\n"
+                f"The closest date for this puzzle type is {nearest_date[0]}."
+            )
+
+        raise XWordDLException("Could not find puzzle for date.")
+
+    def find_solver(self, url):
+        if "puzzles" in url:
+            url = url.removesuffix("/")
+            number = url.split("/")[-1]
+            url = f"https://avxwords.com/download-puzzle/?id={number}"
+        elif "download-puzzle" not in url:
+            raise XWordDLException("Invalid URL for AVCX.")
+
+        try:
+            res = self.session.get(url)
+            res.raise_for_status()
+        except requests.HTTPError:
+            raise XWordDLException(
+                "Unable to open page. Possible authentication issue?"
+            )
+
+        soup = bs(res.text, "lxml")
+
+        difficulty_l = soup.find(
+            "span",
+            class_="difficulty",
+        ).get("aria-label")
+        if difficulty_l:
+            difficulty = difficulty_l.split()[0]
+            self.descriptions.append(f"Difficulty: {difficulty}.")
+
+        desc = soup.find(id="puzzle-newsletter").get_text()
+        if desc:
+            self.descriptions.append(desc.strip())
+
+        al_string = re.compile(".*AcrossLite.*")
+        al_badge = soup.find("span", class_="badge", string=al_string)
+        if al_badge:
+            url = al_badge.parent.get("href")
+            return urljoin(self.base_url, url)
+
+        # if AVCX doesn't have .puz, download JPZ and convert it (TODO)
+        jpz_string = re.compile(".*JPZ.*")
+        jpz_badge = soup.find("span", class_="badge", string=jpz_string)
+        if jpz_badge:
+            url = jpz_badge.parent.get("href")
+            return urljoin(self.base_url, url)
+
+        raise XWordDLException("Could not find valid puzzle file.")
+
+    def fetch_data(self, solver_url):
+        res = self.session.get(solver_url)
+
+        try:
+            res.raise_for_status()
+        except requests.HTTPError:
+            raise XWordDLException("Puzzle data not found.")
+
+        return res.content
+
+    def parse_xword(self, xword_data):
+        # This function can receive either PUZ files or JPZ
+        if xword_data[2:14] == b"ACROSS&DOWN\x00":
+            return self._parse_puz(xword_data)
+        elif xword_data[:5] == b"<?xml":
+            return self._parse_jpz(xword_data)
+        else:
+            raise XWordDLException("Invalid puzzle data.")
+
+    def _parse_puz(self, data):
+        puzzle = puz.load(data)
+        puzzle.author = puzzle.author.strip()
+        puzzle.title = puzzle.title.strip()
+
+        if puzzle.title and ", edited by " in puzzle.title:
+            parts = puzzle.title.partition(", edited by ")
+            puzzle.title = parts[0]
+            self.descriptions.append(f"Edited by {parts[2]}.")
+
+        if self.descriptions:
+            puzzle.notes = "\n\n".join(self.descriptions)
```

Comment:
Per my last comment, I think you could do this like

```
puzzle.notes = "\n\n".join([f"<p>{d}</p>" for d in self.descriptions])
```

Reply:
Okay, I've mostly done that. There's a little more tinkering to get a nice plain text rendering; I'm replacing any links with just the text when `preserve-html` is off, hopefully you think that's a reasonable compromise if no one is actually rendering the Markdown at present.